build.rs

```diff
@@ -0,0 +1,18 @@
+#[cfg(all(feature = "detect-nightly", not(feature = "nightly")))]
+pub fn main() {
+    let rustc_version = std::process::Command::new("rustc")
+        .arg("--version")
+        .output()
+        .expect("Failed to get rustc version")
+        .stdout;
+
+    let rustc_version = String::from_utf8(rustc_version).expect("Failed to parse rustc version");
+    if rustc_version.contains("nightly") {
+        println!("cargo:rustc-cfg=detected_nightly");
+    }
```

Comment:
I don't like the idea of a build failing because an error diagnostics crate cannot detect if it can use nicer diagnostics. Can we get rid of the panics?

Reply:
Thinking about it some more this build script is incorrect. It's plausible `rustc` is not the name of the invoked compiler. I'm not sure what I was thinking, `$RUSTC` is needed.

Reply:
I think I may disagree with removing the panics. 

1. `$RUSTC` or `rustc` doesn't exist / there's a spurious OS error
Rather than build something non-deterministically we should fail fast.

2. `rustc --version` changes
Either it's no longer UTF-8 (not in my rust!) or `"nightly"` no longer appears in _nightly_ toolchains. That seems very unlikely and would break crates like [rustc_version](https://docs.rs/rustc_version/0.4.1/rustc_version/)

Reply:
Reminder: This is not the user's choice, this is a library's choice. If this crate is 5 levels deep in subdeps and it's nightly detection fails in an otherwise passing build, that is a build failure due to a crate wanting to display non-existent errors nicely and requires fixing. That's silly.

I would be okay if you set a config for "failed to nightly detect" and printed a warning alongside existing diagnostics, but that's as far as I am happy going with nightly detection failures.

Reply:
> If you want nightly detection you should get nightly detection.

Sorry, maybe this wasn't the best way to make my point. Can we take a step back? My wonder is what is the specific failure modes you are concerned about.

Reply:
In the sentence "if you want nightly detection", you are talking about a "you". In this case, "you" is not the developer of the binary crate who is affected by build failures, it's the developer of the library depending on proc-macro-error2, and this mismatch is why I am generally against nightly detection in general.

I am not worried about any specific error conditions, I am concerned that nightly detection may lead to an otherwise passing build failing, which is silly for a feature to display nicer error messages.